### DISCUSSION TITLE
[Bug]: story rendering order influences the order of the CSS styles in the head

### DISCUSSION THREAD

**Discussion by martinkutter** [storybookjs/storybook#30925] [Help] [answered]
### Describe the bug

As discovered some years ago in #13515 the name of the story defines in which order the imported styles from the component are added to the head. So if I change the name of a story, or delete the story, or force a rerender the look of a component could change.

### To Reproduce

https://stackblitz.com/edit/github-64tjnp

<img width="166" alt="image" src="https://github.com/storybookjs/storybook/assets/16589094/83c922ab-625d-42e2-92c9-d47975c6237d">

`AHeader` imports `header.module.scss` so it is added first in the head.
`ZHeader` imports `header.module.scss` again, but it's already in the head.
As a result, the button turns green because `button.modules.scss` is added to the head after `header.module.scss`.
<img width="523" alt="image" src="https://github.com/storybookjs/storybook/assets/16589094/8cffa90d-ca0c-41bd-9f34-b42271f842f9">

If I simply resize the window, the order changes and the button turns red.
![Kapture 2023-08-29 at 15 53 09](https://github.com/storybookjs/storybook/assets/16589094/6c52ed57-289d-4b7d-aa67-28b9239f8bfa)

**Comment by dcop** [comment]
This is happening to me also on 7.5.1, but the weird thing is that in dev mode everything works fine, while building and serving is breaking the CSS order. I'm also using CSS modules and my main (css) theme is loaded in the provider that I then use in the `preview` file with `withThemeFromJSXProvider`.

**Comment by lztetreault-dev** [comment]
Also having this issue on `7.5.2`, this seems like it should be a high priority bug. It defeats the point of developing in isolation if your component doesn't look the same as it will in your application.

**Comment by hendrikheil** [comment]
We're on v8.0.0-rc.3 and are encountering the a very similar issue using vite builder. In dev mode the order of the styles is inserted as expected, preview.ts imported (vendor) styles are first, while component or story styles are imported later.

In built-mode, we get the opposite, where preview styles are always last. This causes specificity issues for us in Chromatic snapshots.

Even if we explicitly set `cssCodeSplit: false` the resulting file still has an order we don't expect. We also tried naming the components / stories alphabetically by prefixing them with Z (so they are after p of preview). Nothing worked so far. 

I'm not sure if this is a problem in storybook or vite. But it is causing results we don't expect.

We could fix it by using CSS layers, however Layers don't have enough browser support for us yet, so this isn't a good solution for us.

**Comment by tmeasday** [comment, marked as answer]
I'm not an expert on CSS modules, but in my experience, if you see different rendering behaviour when you access stories/components in a different order, you'll see the same issue in your actual code depending on the order you import those components in your source code. That import order is often driven by the order you visit routes in your app and can be super tricky to track down.

That seems to me to be a bug that you want to fix in your code and will be much easier to debug in SB 🤷 

In the example repro posted (thank you!) I can't quite get the CRA app running in StackBlitz but I downloaded it and added this to the `index.tsx` to simulate the behaviour of a router that does dynamic imports (e.g like NextJS):

```ts
import React from 'react';
import ReactDOM from 'react-dom/client';

const ZHeader = React.lazy(async () => ({ default: (await import('./stories/zHeader')).ZHeader }));
const AHeader = React.lazy(async () => ({ default: (await import('./stories/aHeader')).AHeader }));
const Router = () => {
  if (document.location.search.indexOf('foo') !== -1) {
    return (
      <>
        <AHeader />
        <ZHeader />
      </>
    );
  }

  return <ZHeader />;
};

const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(
  <React.StrictMode>
    <Router />
  </React.StrictMode>
);
```

You can quite clearly see the same behaviour where the `ZHeader` renders green at `/?foo` and red at `/`:

![image](https://github.com/storybookjs/storybook/assets/132554/efd4e9fc-56d8-4a8a-a58b-952f7f27093d)
![image](https://github.com/storybookjs/storybook/assets/132554/73ed515b-5833-4996-95d0-2f54fb717cfb)

**Comment by hendrikheil** [comment]
So while we were working on an unrelated improvement for bundle size optimizations, we changed our upstream library to import each CSS module at the top of every ESM module we emit. This required quite a bit of tinkering with Vite, as it - by default - bundles all CSS and JS into two bundle files in Lib mode, which makes it a lot harder to tree-shake, but doesn't require processing by downstream bundlers.

By making that change, we now only rely on our global styles to provide certain reused design system tokens (as custom properties in :root), but that doesn't depend on any order in our stylesheets.

So for us, these changes made the regression go away, but I still believe storybook should behave the same way in dev and built mode - which it doesn't right now.

**Comment by tmeasday** [comment]
> So for us, these changes made the regression go away, but I still believe storybook should behave the same way in dev and built mode - which it doesn't right now.

Yes, this is a fair point. Do we have any idea of why it behaves differently? In the reproduction @martinkutter posted it broke in dev mode.

**Comment by hendrikheil** [comment]
Not sure really. I did see this issue upstream in vite: https://github.com/vitejs/vite/issues/3924 which might be related. I'm not sure where the root cause is really. I think it'd be interesting to see if the webpack builder exhibits the same behavior as the vite one.

**Comment by kerryj89** [comment]
I am running into the same issue, it seems to happen in published Storybook only. 

In `src/styles/global.css` I have:

```css
@layer theme, base, components, utilities;
```

That's all fine and good as I can see the output in my `preview-*.css` below.  I am using Storybook/Next.js + Vite which uses PostCSS and you can see that while it's not together in one line like above, it breaks it apart in order and at the right time.  This is fine.

![capture-2025-04-01-WCfpUp2A@2x](https://github.com/user-attachments/assets/3c713322-00ae-40c2-8756-38ae905eb90f)

What's not fine is that it injects the global styles at the end.  That makes no sense, and it breaks css layers functionality:
![capture-2025-04-01-yY6xfuaH@2x](https://github.com/user-attachments/assets/5527ee96-12ba-46d1-95dc-cc45503e0442)


Consider a `Tooltip` component that has
```
@layer components {
  .tooltip {
    color: red;
  }
}
```
And it's being placed above the global styles `preview-*.css`:

![capture-2025-04-01-Dup6XjGP](https://github.com/user-attachments/assets/56e1ca40-2968-403f-b69c-231a78132b12)

What actually happens in css land is that it orders things as it comes across it.  This would be fine if `preview-*.css` came first, but instead it reads `@layer components { .tooltip { ... } }` first and so it places it at the beginning of the layers.

So now if you inspect it you will see the order is @layer components, theme, base, utilities.

You can see for yourself - inspect the "hello" text and expand the cascade layers.  If an @layer exists first, it sets it as first and there is no way to change it after:  https://jsfiddle.net/18twz4sc/2/

![capture-2025-04-01-fy9kEpIw](https://github.com/user-attachments/assets/2e187233-7fbb-4e59-a53c-fdca006cd73c)

This is fundamentally broken, more so now with @layer gaining popularity (Tailwind v4 uses it).

--

I am going to see if there is some kind of hack I can use to set the @layer order on the html or one of the storybook html pages in hopes it sees it first before it reaches the css.

--

Edit: Was able to fix this but this will only work if your styles are wrapped in css @layer.  Set your @layer order at this level so that it's seen before the mis-ordered stylesheets.

.storybook/preview-head.html:
```html
<style>
  @layer theme, base, components, utilities;
</style>
```
